tachyon/zk/air/constraint_system/constraint_system.h

```diff
@@ -0,0 +1,140 @@
+#ifndef TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+#define TACHYON_ZK_AIR_CONSTRAINT_SYSTEM_CONSTRAINT_SYSTEM_H_
+
+#include <stddef.h>
+
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "third_party/eigen3/Eigen/Core"
+
+#include "tachyon/zk/air/constraint_system/variable.h"
+#include "tachyon/zk/air/expressions/evaluator.h"
+
+namespace tachyon::zk::air {
+
+template <typename F>
+class ConstraintSystem {
+ public:
+  ConstraintSystem() = default;
+  ConstraintSystem(size_t preprocessed_width, size_t main_width,
+                   size_t num_public_values)
+      : preprocessed_width_(preprocessed_width),
+        main_width_(main_width),
+        num_public_values_(num_public_values) {
+    preprocessed_.reserve(preprocessed_width * 2);
+    main_.reserve(main_width * 2);
+    public_values_.reserve(num_public_values);
+
+    for (size_t row : {0, 1}) {
+      for (Eigen::Index i = 0; i < preprocessed_width_; ++i) {
+        preprocessed_.push_back(
+            Variable::Preprocessed(row * preprocessed_width_ + i));
+      }
+    }
+    for (size_t row : {0, 1}) {
+      for (Eigen::Index i = 0; i < main_width_; ++i) {
+        main_.push_back(Variable::Main(row * main_width_ + i));
+      }
+    }
+    for (size_t i = 0; i < num_public_values; ++i) {
+      public_values_.push_back(Variable::Public(i));
+    }
+  }
+
+  const std::vector<Variable>& GetPublicValues() const {
+    return public_values_;
+  }
+
+  const std::vector<Variable>& GetPreprocessed() const { return preprocessed_; }
+
+  const std::vector<Variable>& GetMain() const { return main_; }
+
+  void EnforceFirstRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::FirstRow(std::move(constraint)));
+  }
+
+  void EnforceLastRowConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::LastRow(std::move(constraint)));
+  }
+
+  void EnforceTransitionConstraint(std::unique_ptr<Expression<F>> constraint) {
+    EnforceConstraint(ExpressionFactory<F>::Transition(std::move(constraint)));
+  }
+
+  void EnforceConstraint(std::unique_ptr<Expression<F>> constraint) {
+    constraints_.push_back(std::move(constraint));
+  }
+
+  template <typename MainDerived>
+  bool IsSatisfied(Evaluator<F>& evaluator, const std::vector<F>& public_values,
+                   std::reference_wrapper<const MainDerived> main) const {
```

Comment:
I'm curious about the reason for intentionally using `std::reference_wrapper<const MainDerived>`, instead of `const MainDerived& main`

Reply:
@dongchangYoo It was leftover from my attempt of removing the template and allowing all matrix types ;). But I finally figured it out and now it has no raw pointers, non-template and still zero-copy. 

Reference: https://eigen.tuxfamily.org/dox/TopicFunctionTakingEigenTypes.html under "How to write generic, but non-templated function?"

Reply:
Sorry it was leftover from my attempt from taking optional matrix as `std::optional<RowMajorMatrix<F>&>`. I cleaned it up and refactored evaluator to remove raw pointer usage.

Reply:
If you need a `RowMajorMatrix<F>` in the `None` state, how about using a pointer?